### PR TITLE
pythonPackages.tpm2-pytss: init at 0.2.4

### PR DESCRIPTION
**Pull request by baloo**
###### Motivation for this change

tpm2-pytss is a python binding on tpm2-tss library

Unlike tpm2-tools, we cant provide a `LD_LIBRARY_PATH` at startup here. even editing the `os.environ['LD_LIBRARY_PATH']` does not work, it's only read at start up.
To overcome this limitation, I patched tpm2-tss itself, to ensure it would only `dlopen` from known path, like any other compiled package would do on nixos.
I think this also solve problems for people who would link against tpm2-tss itself, as I believe they currently would need to provide a `LD_LIBRARY_PATH` which I think is inconvenient.

cc @delroth I'd love if I could get your feedback on this.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/tpm2-tss/default.nix [hunk 1]
```diff
@@ -27,14 +27,40 @@ stdenv.mkDerivation rec {
...
   enableParallelBuilding = true;
 
-  postPatch = "patchShebangs script";
+  patches = [
+    # Do not rely on dynamic loader path
+    # TCTI loader relies on dlopen(), this patch prefixes all calls with the output directory
+    ./no-dynamic-loader-path.patch
+  ];
+  postPatch = ''
+    patchShebangs script
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/tpm2-tss/default.nix:36]:
```suggestion
  ];

  postPatch = ''
    patchShebangs script
```

File: pkgs/development/libraries/tpm2-tss/default.nix [hunk 2]
```diff
@@ -27,14 +27,40 @@ stdenv.mkDerivation rec {
...
+      $out/lib/
+    substituteInPlace ./test/unit/tctildr-dl.c --replace \
+      ', "libtss2' \
+      ", \"$out/lib/libtss2"
+    substituteInPlace ./test/unit/tctildr-dl.c --replace \
+      ', "foo' \
+      ", \"$out/lib/foo"
+    substituteInPlace ./test/unit/tctildr-dl.c --replace \
+      ', TEST_TCTI_NAME' \
+      ", \"$out/lib/\"TEST_TCTI_NAME"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/tpm2-tss/default.nix:48]:
```suggestion
    substituteInPlace src/tss2-tcti/tctildr-dl.c \
      --replace '@PREFIX@' $out/lib/
    substituteInPlace ./test/unit/tctildr-dl.c \
      --replace ', "libtss2' ", \"$out/lib/libtss2"
      --replace ', "foo' ", \"$out/lib/foo" \
      --replace ', TEST_TCTI_NAME' ", \"$out/lib/\"TEST_TCTI_NAME"
```

File: pkgs/development/libraries/tpm2-tss/default.nix [hunk 3]
```diff
@@ -27,14 +27,40 @@ stdenv.mkDerivation rec {
...
     "--enable-integration"
   ];
 
   doCheck = true;
+  preCheck = ''
+    # Since we rewrote the load path in the dynamic loader for the TCTI
+    # The various tcti implementation should be placed in their target directory
+    # before we could run tests
+    installPhase
+    dontInstall=1 # install already done, dont need another one
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/tpm2-tss/default.nix:62]:
```suggestion
    # install already done, dont need another one
    export dontInstall=1
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/tpm2-tss/default.nix (hunk 3 above)
Comment (baloo) [reply pkgs/development/libraries/tpm2-tss/default.nix:62]: The export is not needed as far as I can tell

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 114080` at 2ddbb40d run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>libsForQt512.discover</li>
    <li>libsForQt514.discover</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/bf37ffa303a7813894fd9dbc3a898b8d">fwupd</a></li></li>
    <li><a href="https://gist.github.com/33f21abde1ddd223b024099110295884">gnome-firmware-updater</a></li></li>
    <li><a href="https://gist.github.com/15059f87fe5b52db76493d9b0df00f9e">libsForQt5.discover</a></li> (libsForQt515.discover ,plasma5Packages.discover)</li>
    <li><a href="https://gist.github.com/0ab57b78f3cbc61cb6357717578b66fe">python37Packages.tpm2-pytss</a></li></li>
    <li><a href="https://gist.github.com/5c71b22171f6cae4951b45ec388b018b">python38Packages.tpm2-pytss</a></li></li>
    <li><a href="https://gist.github.com/7a18ffdabc4e5d1e10a74d839d4e65e0">python39Packages.tpm2-pytss</a></li></li>
    <li><a href="https://gist.github.com/21d63f382458d2c83d83a9ed50b5f506">tpm2-abrmd</a></li></li>
    <li><a href="https://gist.github.com/22724e328457fa85c527c9e44c7fb740">tpm2-pkcs11</a></li></li>
    <li><a href="https://gist.github.com/9d48f945fe85439cce88c61feef47444">tpm2-tools</a></li></li>
    <li><a href="https://gist.github.com/db045b57e9d86523130e0a0a37777389">tpm2-tss</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `stdenv`.
Near pkgs/development/python-modules/tpm2-pytss/default.nix:1:8:
```
  |
1 | { lib, stdenv, buildPythonPackage, fetchPypi, pythonOlder
  |        ^
```
</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 114080` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>libsForQt512.discover</li>
<li>libsForQt514.discover</li>
</ul>
</details>
<details>
<summary>
11 packages built:
</summary>
<ul>
<li>fwupd</li>
<li>gnome-firmware-updater</li>
<li>gnome3.gnome-software</li>
<li>libsForQt5.discover (libsForQt515.discover ,plasma5Packages.discover)</li>
<li>python37Packages.tpm2-pytss</li>
<li>python38Packages.tpm2-pytss</li>
<li>python39Packages.tpm2-pytss</li>
<li>tpm2-abrmd</li>
<li>tpm2-pkcs11</li>
<li>tpm2-tools</li>
<li>tpm2-tss</li>
</ul>
</details>

**Pull request merged**